staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,370 @@
+		if err != nil {
+			klog.ErrorS(err, "error getting ip and port info of the remote server while proxying")
+			responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Error getting ip and port info of the remote server while proxying"), s, gv, w, r)
+			return
+		}
+
+		h.proxyRequestToDestinationAPIServer(r, w, destServerIP)
+
+	})
+}
+
+func (h *peerProxyHandler) findServiceableByServers(gvr schema.GroupVersionResource, localAPIServerId string, reconciler reconcilers.PeerEndpointLeaseReconciler) (serviceableByResponse, error) {
+
+	apiserversi, ok := h.svMap.Load(gvr)
+
+	// no value found for the requested gvr in svMap
+	if !ok || apiserversi == nil {
+		return serviceableByResponse{}, fmt.Errorf("no StorageVersions found for the GVR: %v", gvr)
+	}
+	apiservers := apiserversi.(*sync.Map)
+	response := serviceableByResponse{}
+	var peerServerIPs []string
+	apiservers.Range(func(key, value interface{}) bool {
+		apiserverKey := key.(string)
+		if apiserverKey == localAPIServerId {
+			response.locallyServiceable = true
+			// stop iteration
+			return false
+		}
+
+		ip, err := reconciler.GetEndpoint(apiserverKey)
+		if err != nil {
+			// continue with iteration
+			klog.Errorf("failed to get peer ip from storage lease for server %s", apiserverKey)
+			return true
+		}
+		peerServerIPs = append(peerServerIPs, ip)
+		// continue with iteration
+		return true
+	})
+	response.peerIPs = peerServerIPs
+	return response, nil
+}
+
+func (h *peerProxyHandler) proxyRequestToDestinationAPIServer(req *http.Request, rw http.ResponseWriter, host string) {
+	user, ok := apirequest.UserFrom(req.Context())
+	if !ok {
+		klog.Errorf("failed to get user info from request")
+		return
+	}
+
+	// write a new location based on the existing request pointed at the target service
+	location := &url.URL{}
+	location.Scheme = "https"
+	location.Host = host
+	location.Path = req.URL.Path
+	location.RawQuery = req.URL.Query().Encode()
+
+	newReq, cancelFn := apiserverproxyutil.NewRequestForProxy(location, req)
+	newReq.Header.Add(PeerProxiedHeader, "true")
+	defer cancelFn()
+
+	// create transport
+	clientConfig := &transport.Config{
+		TLS: transport.TLSConfig{
+			Insecure:   false,
+			CertFile:   h.proxyClientCertFile,
+			KeyFile:    h.proxyClientKeyFile,
+			CAFile:     h.peerCAFile,
+			ServerName: "kubernetes.default.svc",
+		},
+	}
+
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	defer cancelCtx()
+	proxyRoundTripper, transportBuildingError := transport.New(clientConfig)
+	if transportBuildingError != nil {
+		klog.Error(transportBuildingError.Error())
+		return
+	}
+
+	proxyRoundTripper = transport.NewAuthProxyRoundTripper(user.GetName(), user.GetGroups(), nil, proxyRoundTripper)
+
+	delegate := &epmetrics.ResponseWriterDelegator{ResponseWriter: rw}
+	w := responsewriter.WrapForHTTP1Or2(delegate)
+
+	handler := proxy.NewUpgradeAwareHandler(location, proxyRoundTripper, true, false, &responder{w: w, ctx: req.Context()})
```

Comment:
do we need to use the upgradeAwareHandler here @liggitt ? it sounds to me like we should not , since this should only handle the resources not present

If that is the case doing ` handler  := httputil.NewSingleHostReverseProxy(remote)`  should be enough